analyzer/src/main/kotlin/managers/Yarn2.kt

```diff
@@ -0,0 +1,127 @@
+
+import java.io.File
+
+import org.ossreviewtoolkit.analyzer.AbstractPackageManagerFactory
+import org.ossreviewtoolkit.analyzer.managers.utils.hasYarnLockFile
+import org.ossreviewtoolkit.analyzer.managers.utils.mapDefinitionFilesForYarn2
+import org.ossreviewtoolkit.model.config.AnalyzerConfiguration
+import org.ossreviewtoolkit.model.config.RepositoryConfiguration
+import org.ossreviewtoolkit.model.jsonMapper
+import org.ossreviewtoolkit.model.yamlMapper
+import org.ossreviewtoolkit.utils.common.ProcessCapture
+
+/**
+ * The [Yarn2+](https://next.yarnpkg.com/) package manager for JavaScript.
+ */
+class Yarn2(
+    name: String,
+    analysisRoot: File,
+    analyzerConfig: AnalyzerConfiguration,
+    repoConfig: RepositoryConfiguration
+) : Npm(name, analysisRoot, analyzerConfig, repoConfig) {
+    class Factory : AbstractPackageManagerFactory<Yarn2>("Yarn2") {
+        override val globsForDefinitionFiles = listOf("package.json")
+
+        override fun create(
+            analysisRoot: File,
+            analyzerConfig: AnalyzerConfiguration,
+            repoConfig: RepositoryConfiguration
+        ) = Yarn2(managerName, analysisRoot, analyzerConfig, repoConfig)
+    }
+
+    /**
+     * The Yarn2+ executable is not available in ORT's Docker image: we use the program shipped in .yarn/releases.
+     * The resource file `.yarnrc.yml` contains the path to the executable under the property name `yarnPath` defined
+     * for current project e.g. yarnPath: .yarn/releases/yarn-3.2.1.cjs.
+     * This map holds the mapping between the directory and their Yarn2 executables.
+     */
+    private val yarn2CommandsByPath: MutableMap<File, String> = mutableMapOf()
+
+    /**
+     * A NPM package manager, used to call NPM when Yarn2+ does not provide enough information.
+     */
+    private val delegateNpmAnalyzer = Npm("npm", analysisRoot, analyzerConfig, repoConfig)
+
+    companion object {
+        const val YARN2_RESOURCE_FILE = ".yarnrc.yml"
+
+        // The property in `.yarnrc.yml`containing the path to the Yarn 2+ executable.
+        const val YARN_PATH_PROPERTY_NAME = "yarnPath"
+    }
+
+    override fun hasLockFile(projectDir: File) = hasYarnLockFile(projectDir)
+
+    override fun command(workingDir: File?): String {
+        return workingDir?.let {
+            yarn2CommandsByPath.getOrPut(it) {
+                val yarnConfig = yamlMapper.readTree(workingDir.resolve(YARN2_RESOURCE_FILE))
+                val yarnCommand = yarnConfig.get(YARN_PATH_PROPERTY_NAME)
+                yarnCommand?.let { command ->
+                    val yarnExecutable = workingDir.resolve(command.textValue())
+                    // TODO: Yarn2 executable is a `cjs` file. Check if under Windows it needs to be run with `node`.
+
+                    // TODO: This is a security risk to blindly run code coming from a repository other than ORT's. ORT
+                    //  should download the Yarn2 binary from the official repository and run it.
```

Comment:
Just an idea: Maybe it is possible to check for the hash to ensure that the executable is valid?

Reply:
This is a gook idea but I don't know where to find the reference hash.

They don't put the hashes on their Github:
https://github.com/yarnpkg/berry/releases

They have no hash nor index page on the repo e.g. 
https://repo.yarnpkg.com/3.1.0/packages/yarnpkg-cli/bin/yarn.js